src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -0,0 +1,419 @@
+using System.Threading.Tasks;
+using static System.Net.Quic.Implementations.MsQuic.Internal.MsQuicNativeMethods;
+
+namespace System.Net.Quic.Implementations.MsQuic
+{
+    internal sealed class MsQuicConnection : QuicConnectionProvider
+    {
+        // Functions to invoke in MsQuic
+        public MsQuicApi _api;
+        private MsQuicSession _session;
+
+        // Pointer to the underlying connection
+        private IntPtr _ptr;
+
+        // Handle to this object for native callbacks.
+        private GCHandle _handle;
+
+        // Delegate that wraps the static function that will be called when receiving an event.
+        private ConnectionCallbackDelegate _connectionDelegate;
+
+        // Endpoint to either connect to or the endpoint already accepted.
+        private IPEndPoint _localEndPoint;
+        private readonly IPEndPoint _remoteEndPoint;
+
+        private readonly ResettableCompletionSource<uint> _connectTcs = new ResettableCompletionSource<uint>();
+        private readonly ResettableCompletionSource<uint> _shutdownTcs = new ResettableCompletionSource<uint>();
+
+        private bool _disposed;
+        private bool _connected;
+
+        // Queue for accepted streams
+        private readonly Channel<MsQuicStream> _acceptQueue = Channel.CreateUnbounded<MsQuicStream>(new UnboundedChannelOptions()
+        {
+            SingleReader = true,
+            SingleWriter = true,
+        });
+
+        // constructor for inbound connections
+        public MsQuicConnection(IPEndPoint localEndPoint, IPEndPoint remoteEndPoint, MsQuicApi api, IntPtr nativeObjPtr)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+            _localEndPoint = localEndPoint;
+            _remoteEndPoint = remoteEndPoint;
+            _api = api;
+            _ptr = nativeObjPtr;
+
+            SetCallbackHandler();
+            SetIdleTimeout(TimeSpan.FromSeconds(120));
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+        }
+
+        // constructor for outbound connections
+        public MsQuicConnection(QuicClientConnectionOptions options)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            // TODO need to figure out if/how we want to expose sessions
+            // Creating a session per connection isn't ideal.
+            _session = new MsQuicSession();
+            _ptr = _session.ConnectionOpen(options);
+            _remoteEndPoint = options.RemoteEndPoint;
+            _api = MsQuicApi.Api;
+
+            SetCallbackHandler();
+            SetIdleTimeout(TimeSpan.FromSeconds(120));
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+        }
+
+        internal override IPEndPoint LocalEndPoint
+        {
+            get
+            {
+                if (!_connected)
+                {
+                    throw new InvalidOperationException("Connection must be started before getting endpoint.");
```

Comment:
This isn't always true, right? You can specify a particular local endpoint in the constructor.

Reply:
And even if you don't, should we actually throw here? Or return null? Not sure...

Reply:
Yeah, this is an interesting case. If someone specifies port 0, this wouldn't be resolved until the connection is connected. Maybe we have to check for port 0 explicitly here OR we just return the endpoint with port 0 before it's connected and then return the real one afterwards?

Reply:
We should probably just check what Socket does and then do the same thing here. Unless what Socket does is really weird, which is possible.

Reply:
I started to try to understand what LocalEndpoint was doing in Socket.cs... gave up after a good minute 😄 https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L190

Reply:
Yeah, it's not pleasant. Honestly I'd just write some code that uses Socket and see what it does.

Reply:
The call to Connect takes in the EndPoint rather than the constructor. Before that, the LocalEndPoint is null.

Maybe we can consider changing when we pass the endpoint in (move it to connect).